junit-jupiter-params/src/main/java/org/junit/jupiter/params/converter/DefaultArgumentConverter.java

```diff
@@ -0,0 +1,171 @@
+		if (input == null) {
+			if (targetClass.isPrimitive()) {
+				throw new ArgumentConversionException(
+					"Cannot convert null to primitive value of type " + targetClass.getName());
+			}
+			return null;
+		}
+		return convertToReferenceType(input, toWrapperType(targetClass));
+	}
+
+	private Class<?> toWrapperType(Class<?> targetClass) {
+		Class<?> wrapperType = ReflectionUtils.getWrapperType(targetClass);
+		return wrapperType != null ? wrapperType : targetClass;
+	}
+
+	private Object convertToReferenceType(Object input, Class<?> targetClass) {
+		if (targetClass.isInstance(input)) {
+			return input;
+		}
+		if (input instanceof String) {
+			Optional<StringConversion> conversion = stringConversions.stream().filter(
+				candidate -> candidate.isResponsible(targetClass)).findFirst();
+			if (conversion.isPresent()) {
+				try {
+					return conversion.get().convert((String) input, targetClass);
+				}
+				catch (Exception ex) {
+					throw new ArgumentConversionException(
+						"Error converting String to type " + targetClass.getName() + ": " + input, ex);
+				}
+			}
+		}
+		throw new ArgumentConversionException("No implicit conversion to convert object of type "
+				+ input.getClass().getName() + " to type " + targetClass.getName());
+	}
+
+	interface StringConversion {
+
+		boolean isResponsible(Class<?> targetClass);
+
+		Object convert(String input, Class<?> targetClass) throws Exception;
+
+	}
+
+	static class PrimitiveStringConversion implements StringConversion {
+
+		private static final Map<Class<?>, Function<String, ?>> CONVERTERS;
+		static {
+			Map<Class<?>, Function<String, ?>> converters = new HashMap<>();
+			converters.put(Boolean.class, Boolean::valueOf);
+			converters.put(Character.class, input -> {
+				Preconditions.condition(input.length() == 1, () -> "String must have length of 1: " + input);
+				return input.charAt(0);
+			});
+			converters.put(Byte.class, Byte::valueOf);
+			converters.put(Short.class, Short::valueOf);
+			converters.put(Integer.class, Integer::valueOf);
+			converters.put(Long.class, Long::valueOf);
+			converters.put(Float.class, Float::valueOf);
+			converters.put(Double.class, Double::valueOf);
+			CONVERTERS = unmodifiableMap(converters);
+		}
+
+		@Override
+		public boolean isResponsible(Class<?> targetClass) {
+			return CONVERTERS.containsKey(targetClass);
+		}
+
+		@Override
+		public Object convert(String input, Class<?> targetClass) {
+			return CONVERTERS.get(targetClass).apply(input);
+		}
+	}
+
+	static class EnumStringConversion implements StringConversion {
+
+		@Override
+		public boolean isResponsible(Class<?> targetClass) {
+			return targetClass.isEnum();
+		}
+
+		@Override
+		public Object convert(String input, Class<?> targetClass) throws Exception {
+			return valueOf(targetClass, input);
+		}
+
+		@SuppressWarnings({ "unchecked", "rawtypes" })
+		private Object valueOf(Class targetClass, String input) {
+			return Enum.valueOf(targetClass, input);
```

Comment:
Yep, you're right: `convertToTargetType` catches the exception.

And you're also right: tests wouldn't be bad. 😉